examples/serialize_to_disk.rs

```diff
@@ -0,0 +1,172 @@
+//! a save mechanism for a game.
+//!
+//! The example creates a sample `World`, serializes it, saves it to disk as `saved_world.world` file,
+//! loads it back from the disk, deserializes the loaded world data and validates that component
+//! data of the worlds match.
+//!
+//! Run this example from crate root with:
+//! `cargo run --example serialize_to_disk --features "column-serialize"`
+
+use hecs::serialize::column::{
+    deserialize_column, try_serialize, try_serialize_id, DeserializeContext, SerializeContext,
+};
+use hecs::{Archetype, ColumnBatchBuilder, ColumnBatchType, World};
+use serde::{Deserialize, Serialize, Serializer};
+use std::any::TypeId;
+use std::fs::File;
+use std::io::{BufReader, Write};
+use std::path::Path;
+
+// Identifiers for the components we want to include in the serialization process:
+#[derive(Serialize, Deserialize)]
+enum ComponentId {
+    ComponentA,
+    ComponentB,
+}
+
+// We need to implement a context type for the hecs serialization process:
+#[derive(Default)]
+pub struct SaveContext {
+    pub components: Vec<ComponentId>,
+}
+
+// Components of our world.
+// Only Serialize and Deserialize derives are necessary.
+#[derive(Serialize, Deserialize, Eq, PartialEq, Debug, Clone)]
+pub struct ComponentA {
+    pub data: usize,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Debug, Clone)]
+pub struct ComponentB {
+    pub some_other_data: String,
+}
+
+#[cfg(feature = "column-serialize")]
+impl DeserializeContext for SaveContext {
+    fn deserialize_component_ids<'de, A>(&mut self, mut seq: A) -> Result<ColumnBatchType, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        self.components.clear(); // Discard data from the previous archetype
+        let mut batch = ColumnBatchType::new();
+        while let Some(id) = seq.next_element()? {
+            match id {
+                ComponentId::ComponentA => {
+                    batch.add::<ComponentA>();
+                }
+                ComponentId::ComponentB => {
+                    batch.add::<ComponentB>();
+                }
+            }
+            self.components.push(id);
+        }
+        Ok(batch)
+    }
+
+    fn deserialize_components<'de, A>(
+        &mut self,
+        entity_count: u32,
+        mut seq: A,
+        batch: &mut ColumnBatchBuilder,
+    ) -> Result<(), A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        // Decode component data in the order that the component IDs appeared
+        for component in &self.components {
+            match *component {
+                ComponentId::ComponentA => {
+                    deserialize_column::<ComponentA, _>(entity_count, &mut seq, batch)?;
+                }
+                ComponentId::ComponentB => {
+                    deserialize_column::<ComponentB, _>(entity_count, &mut seq, batch)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl SerializeContext for SaveContext {
```

Comment:
This shouldn't be implemented on the same type as `DeserializeContext`, because it doesn't need the same (or any) data on the side.

Reply:
So should it be done like this?

```rust
#[derive(Default)]
pub struct SaveContextSerialize {}
#[derive(Default)]
pub struct SaveContextDeserialize {
   pub components: Vec<ComponentId>,
}
```

Reply:
That would be clearer, yes.